bindings/experimental/distrdf/python/DistRDF/ComputationGraphGenerator.py

```diff
@@ -106,6 +106,83 @@ def _(operation: Snapshot, range_id: int) -> None:
+    in order to bind their lifetimes and also have an input RDataFrame node for
+    the next recursive state.
+    """
+    future_results = []
+
+    rdf_operation = getattr(previous_node, distrdf_node.operation.name)
+    _make_op_lazy_if_needed(distrdf_node.operation, range_id)
+    pyroot_node = rdf_operation(*distrdf_node.operation.args, **distrdf_node.operation.kwargs)
+
+    # The result is a PyROOT object which is stored together with
+    # the DistRDF node. This binds the PyROOT object lifetime to the
+    # DistRDF node, so both nodes will be kept alive as long as there
+    # is a valid reference pointing to the DistRDF node.
+    distrdf_node.pyroot_node = pyroot_node
+
+    append_node_to_results(distrdf_node.operation, pyroot_node, future_results)
+    return future_results
+
+
+@_get_results_list.register
+def _(distrdf_node: HeadNode, previous_node: Any, range_id: int) -> List:
+    """
+    Implementation of the initial state of the computation_graph_generator
+    function. The 'previous_node' parameter is some kind of ROOT::RDataFrame.
+    The lifetimes of the DistRDF head node and its RDataFrame counterpart are
+    bound together, in order to provide an input for the next recursive state.
+    """
+    distrdf_node.pyroot_node = previous_node
+    return []
+
+
+def generate_computation_graph(distrdf_node: Node, previous_node: Any, range_id: int) -> List:
+    """
+    Generates the RDF computation graph by recursively retrieving
+    information from the DistRDF nodes.
+
+    Args:
+        distrdf_node: The current DistRDF node in
+            the computation graph. In the first recursive state this is None
+            and it will be set equal to the DistRDF headnode.
+        previous_node: The node in the RDF computation graph on which
+            the operation of the current recursive state is called. In the
+            first recursive state, this corresponds to the RDataFrame
+            object that will be processed. Specifically, if the head node
+            of the computation graph is an EmptySourceHeadNode, then the
+            first current node will actually be the result of a call to the
+            Range operation. If the head node is a TreeHeadNode then the
+            node will be an actual RDataFrame. Successive recursive states
+            will receive the result of an RDF operation call
+            (e.g. Histo1D, Count).
+        range_id: The id of the current range. Needed to assign a
+            file name to a partial Snapshot if it was requested.
+
+    Returns:
+        list: List of actions of the computation graph to be triggered. Each
+        element is some kind of promise of a result (usually an
+        RResultPtr). Exceptions are the 'AsNumpy' operation for which an
+        'AsNumpyResult' is returned and the 'Snapshot' operation for which a
+        'SnapshotResult' is returned.
+    """
+    future_results = _get_results_list(distrdf_node, previous_node, range_id)
```

Comment:
Should this be called `_run_operation` or similar, instead of `_get_results_list`, to indicate better what the helper function is doing (which is, except for the head node, running the current operation on the real RDF)?